### PR TITLE
[SPARK-21533][SQL] Print warning messages when override function `configure` found in Hive UDFs

### PR DESCRIPTION
**Pull request by maropu**
## What changes were proposed in this pull request?
This pr added code to print warning messages when an unsupported function `public void configure(MapredContext mapredContext)` used in `GenericUDF` and `GenericUDF`. This is because Spark does not call `configure` internally and this is error-prone.

## How was this patch tested?
Manually checked. If this pr applied, you hit a warning message below;
```
scala> sql("CREATE TEMPORARY FUNCTION test AS 'test.TestUDTF'")
17/07/29 11:57:08 WARN HiveShim: Found an overridden method `configure` in TestUDTF, but Spark does not call the method during initialization because Spark does not use MapredContext inside (See SPARK-21533). So, you might reconsider the implementation of TestUDTF.
```

### CODE REVIEW

**Comment by maropu** [comment]
@gatorsmile @hvanhovell What do you think this? Thanks!

**Comment by maropu** [comment]
Jenkins, retest this please.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala [hunk 1]
```diff
@@ -111,6 +112,31 @@ private[hive] object HiveShim {
...
+    if (isSubClassOf(clazz, clsTag)) {
+      val funcClass = clazz.getMethod(funcName, classOf[MapredContext])
+      funcClass.getDeclaringClass != clsTag
+    } else {
+      false
+    }
+  }
+
+  def validateHiveUserDefinedFunction(udfClass: Class[_]): Unit = {
+    if (hasInheritanceOf[GenericUDF]("configure", udfClass) ||
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala]: When `GenericUDF` API has `configure` method? Seems `GenericUDF` at 0.10.0 has no such method?

https://hive.apache.org/javadocs/r0.10.0/api/org/apache/hadoop/hive/ql/udf/generic/GenericUDF.html

Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala]: I checked it seemed this API had been implemented in v0.11.0 (https://github.com/apache/hive/blob/release-0.11.0/ql/src/java/org/apache/hadoop/hive/ql/udf/generic/GenericUDF.java). This is [the commit](https://github.com/apache/hive/commit/a4705e51809e0f6c24cdc417c91600627cbe3482#diff-3111a4cd48c72a32f7713451d7b9ca1aR98) for this API.

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala]: If ran with a Hive version that `configure` is not implemented yet, is `hasInheritanceOf` safe from `NoSuchMethodException `?

Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala]: IIUC spark always refers to `hive-exec-1.2.1.spark2.jar `, so it seems we have no chance to get the exception. But, I think it is not a bad idea to catch `NoSuchMethodException` there for understandability.

Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala]: Oh, that's right. As you said, it is still good to catch the exception.

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala [hunk 2]
```diff
@@ -111,6 +113,35 @@ private[hive] object HiveShim {
     }
   }
 
+  private def isSubClassOf(t: Type, parent: Class[_]): Boolean = t match {
+    case cls: Class[_] => parent.isAssignableFrom(cls)
```
Comment (viirya) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala]: Will you pass in something not a `Class[_]`? If not, we can simply inline `isAssignableFrom` check into `hasInheritanceOf`.

Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveShim.scala]: ok

**Comment by viirya** [comment]
Btw, is it possibly to add an unit test for it?

**Comment by gatorsmile** [comment]
Is this the only function we miss for different types of Hive UD(A/T)Fs?

**Comment by gatorsmile** [comment]
Could we document this limit in the https://spark.apache.org/docs/latest/sql-programming-guide.html#unsupported-hive-functionality?

File: sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala [hunk 3]
```diff
@@ -66,6 +66,8 @@ private[sql] class HiveSessionCatalog(
    * Construct a [[FunctionBuilder]] based on the provided class that represents a function.
    */
   private def makeFunctionBuilder(name: String, clazz: Class[_]): FunctionBuilder = {
+    validateHiveUserDefinedFunction(clazz)
```
Comment (gatorsmile) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: I am neural to introduce a warning message for this case. Not sure how helpful the warning message will be.

Comment (maropu) [inline sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveSessionCatalog.scala]: ok. So, I think it's ok to revisit this again if we have more reports from users.

**Comment by maropu** [comment]
@viirya I've no idea about a way to add tests for this kind of cases and I think the previous prs similar to this case (just add code for warning) also had no test. But, if we could, we'd better to do...

@gatorsmile I checked these Hive classes and I found some other functions that Spark currently ignores... (e.g., `GenericUDF.getRequiredJars`, `GenericUDF.close`...). Also, Spark currently invokes the deprecated `initialize`  interface only in `GenericUDTF`, so users get stuck if they use the non-deprecated `initialize` interface. (See: https://github.com/apache/spark/pull/18527). As you suggested above, it seems we better leave some documents about these unsupported behaviours.

**Comment by gatorsmile** [comment]
Yes! Thanks for your efforts! Please update the documentation at first.

**Comment by maropu** [comment]
ok, I'll make a pr. Thanks!

**Comment by maropu** [comment]
I think the document about these unsupported functions is enough for users, so I'll close this for now. If we have more response from users, we could revisit this again. Thanks.

**Pull request closed without merging**